src/util/expr_util.cpp

```diff
@@ -201,3 +202,65 @@ if_exprt lift_if(const exprt &src, std::size_t operand_number)
 
   return result;
 }
+
+bool is_constantt::is_constant(const exprt &expr) const
+{
+  if(expr.is_constant())
+    return true;
+
+  if(expr.id() == ID_address_of)
+  {
+    const address_of_exprt &address_of_expr = to_address_of_expr(expr);
+
+    return is_constant_address_of(address_of_expr.object());
+  }
+  else if(expr.id() == ID_typecast)
+  {
+    const typecast_exprt &typecast_expr = to_typecast_expr(expr);
+
+    return is_constant(typecast_expr.op());
+  }
+  else if(expr.id() == ID_array_of)
+  {
+    return is_constant(expr.op0());
+  }
+  else if(
+    expr.id() == ID_plus || expr.id() == ID_mult || expr.id() == ID_array ||
```

Comment:
What about divide, floating-point ops, etc.  Do we have any exprt's that cannot be constant folded if all arguments are constant?  If we do; should we?

I realise that constant folding and expression forwarding in symex is more involved than just "is it constant" but...

Reply:
#3095 for further discussion.